### ISSUE TITLE
spec(deployment): the deployment system

### ISSUE THREAD

**Issue by douglasduteil** [douglasduteil/angular-utility-belt#11]
<br>

<div align="center">
<h1>Deployment system</h1>
</div>

<br>

The component deployment is another independent key feature of the system.
I was making a [publishing system for angular-ui](https://github.com/angular-ui/angular-ui-publisher). In short it was making Travis CI the publisher of all releases based on source tagging. However it's already getting old and have various problems concerning the inner template customization and, recently, the sync of the pushed gh-pages on Github for a deploy key.

So for me the way to go is making more atomic module to play with.
Here is my wishlist :
- [ ] **Bumping system** : deal with `package.json`, `bower.json` or `*.json` file with a `version` key.
- [ ] **Git pusher** : will push the code for you based on the `package.json`, `bower.json` or `*.json` file with a `version` key. 
  - Can push tags.  
  - Can push to a _orphan_ branch
  - Can push only the content of a certain directory (like only the content of the `dist` directory)

--- [inspred by angular-ui-publisher#publish.js](https://github.com/angular-ui/angular-ui-publisher/blob/master/lib/publish.js)
<br>

For me the best result is :
- [ ] No distribution files in the sources (master branch).
- [ ] A branch with it's own history (_orphan_ branch) for the distribution files.
- [ ] Always double tagging on releases. Tag for the sources on the master branch (src1.2.3) for the release on the distribution branch (v1.2.3).
- [ ] The distribution branch must content : 
  - [ ] The rooted distribution files
  - [ ] A `package.json`, `bower.json` or `*.json` file to expose the _version_ etc...with no _devdependencies_.
  - [ ] generated CHANGELOG and README

--- [inspired by ui-ace#bower](https://github.com/angular-ui/ui-ace/tree/bower)
<br>

What you think guys @mgcrea @revolunet

**Comment by revolunet** [comment]
100% agree with having a separate dist branch with tags, usable with bower.
Changelog + Readme also in the src of course

Does the bump automatically increment the minor version ?
Can the user force the version number (say this is a major change) ?

**Comment by douglasduteil** [comment]
> Does the bump automatically increment the minor version ?
> Can the user force the version number (say this is a major change) ?

Yes it's in the "Bumping system" for me I was thinking about using https://github.com/ikari-pl/gulp-tag-version maybe

**Comment by mgcrea** [comment]
@douglasduteil, something like [vojtajina/grunt-bump](https://github.com/vojtajina/grunt-bump) (might worth checking: [gulp-bump](https://github.com/stevelacy/gulp-bump)) would be nice indeed.

BTW, @revolunet tu me parlais de mercredi soir dans ton texto pour bosser sur ngFactory, on avait prévu quelque chose (demain du coup)?!

**Comment by douglasduteil** [comment]
Need to be forked not ?https://github.com/boennemann/grunt-semantic-release

**Comment by douglasduteil** [comment]
From :http://blog.hood.ie/2014/09/semantic-releases/

**Comment by mgcrea** [comment]
@douglasduteil nice pick, not sure about the fork, it may be faster to just restart from scratch, there is grunt stuff all over the place. Maybe we should create an issue about changing it to be build-tool agnostic and see if the author is interested.

**Comment by douglasduteil** [comment]
@mgcrea Agreed. I'm just starting getting how it's works....

**Comment by douglasduteil** [comment]
### Pleasantly surprised

My [dd-deploy](https://github.com/douglasduteil/dd-deploy) is working with no modification for git commiting, tagging, and pushing.

I'm making a poke to connect is to Travis CI (for release publishing). And it works !
The _setting up_  process is still to clarify and simplify. For now, the only requirement to make it work with TravisCI is by using :
- **a deploy key** :  Can be generated by any collaborator and placed in the `~/.ssh` folder.
  Pro : repo specific.
  Cons : A hell to setup ( see https://gist.github.com/douglasduteil/5525750#file-travis-secure-key-sh ).
- **a github oauth token** : Can be generated from any profile setting and placed in the repo remote URL as `https://${GITHUB_TOKEN}@github.com/<user>/<repo>`.
  Pro : less to de/encode. Just a Travis env to set. 
  Cons : global access to all the repo.

**Comment by douglasduteil** [comment]
S'up boys

I gat my stuff working :)
check the _deployor_ I made : https://github.com/douglasduteil/node-git-deployor

**Comment by revolunet** [comment]
woot, looks cool :)
poke = POC ?

the deploy key is the way to go. looks like we can replace the gem install with this : https://github.com/pwmckenna/node-travis-encrypt ?

**Comment by mgcrea** [comment]
Congratz @douglasduteil ! I'll check that this week!

On my side, I've worked on the gulp workflow this weekend and found some really interesting modules to refactor everything: [lazypipe](https://github.com/OverZealous/lazypipe) + [gulp-inject](https://github.com/klei/gulp-inject), that enables to reuse any combination of transforms, and reinject this combination during the index.{jade,html} processing, it does really simplify the whole gulp file/process. I'll retrofit this into ngFactory very soon!

**Comment by douglasduteil** [comment]
:+1:
